pkg/inclusion/nmt_caching_test.go

```diff
@@ -0,0 +1,178 @@
+package inclusion
+
+import (
+	"testing"
+
+	"github.com/celestiaorg/celestia-app/testutil/coretestutil"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+)
+
+func TestWalkCachedSubTreeRoot(t *testing.T) {
+	// create the first main tree
+	strc := newSubTreeRootCacher()
+	oss := uint64(8)
+	tr := wrapper.NewErasuredNamespacedMerkleTree(oss, nmt.NodeVisitor(strc.Visit))
+	d := []byte{0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 3, 4, 5, 6, 7, 8}
+	for i := 0; i < 8; i++ {
+		tr.Push(d, rsmt2d.SquareIndex{
+			Axis: uint(rsmt2d.Row),
+			Cell: uint(i),
+		})
+	}
+	highestRoot := tr.Root()
+
+	// create a short sub tree
+	shortSubTree := wrapper.NewErasuredNamespacedMerkleTree(oss)
+	for i := 0; i < 2; i++ {
+		shortSubTree.Push(d, rsmt2d.SquareIndex{
+			Axis: uint(rsmt2d.Row),
+			Cell: uint(i),
+		})
+	}
+	shortSTR := shortSubTree.Root()
+
+	// create a tall sub tree root
+	tallSubTree := wrapper.NewErasuredNamespacedMerkleTree(oss)
+	for i := 0; i < 4; i++ {
+		tallSubTree.Push(d, rsmt2d.SquareIndex{
+			Axis: uint(rsmt2d.Row),
+			Cell: uint(i),
+		})
+	}
+	tallSTR := tallSubTree.Root()
+
+	type test struct {
+		name        string
+		path        []bool
+		subTreeRoot []byte
+		error       string
+	}
+
+	tests := []test{
+		{
+			"left most short sub tree",
+			[]bool{false, false},
+			shortSTR,
+			"",
+		},
+		{
+			"left middle short sub tree",
+			[]bool{false, true},
+			shortSTR,
+			"",
+		},
+		{
+			"right middle short sub tree",
+			[]bool{true, false},
+			shortSTR,
+			"",
+		},
+		{
+			"right most short sub tree",
+			[]bool{true, true},
+			shortSTR,
+			"",
+		},
+		{
+			"left most tall sub tree",
+			[]bool{false},
+			tallSTR,
+			"",
+		},
+		{
+			"right most tall sub tree",
+			[]bool{true},
+			tallSTR,
+			"",
+		},
```

Comment:
Shouldn't we add also cases where this would fail? Like for an empty path and a path longer than the depth of the tree?

Reply:
good idea! added a case here https://github.com/celestiaorg/celestia-app/pull/549/commits/b6745f6f6acdfbd724898610096bf32c7cb3e088